core/stateless/witness.go

```diff
@@ -76,11 +73,8 @@ func NewWitness(chain HeaderReader, block *types.Block) (*Witness, error) {
 // the chain to cover the block being added.
 func (w *Witness) AddBlockHash(number uint64) {
 	// Keep pulling in headers until this hash is populated
-	for int(w.Block.NumberU64()-number) > len(w.Headers) {
-		tail := w.Block.Header()
-		if len(w.Headers) > 0 {
-			tail = w.Headers[len(w.Headers)-1]
-		}
+	for int(w.context.Number.Uint64()-number) > len(w.Headers) {
```

Comment:
Theoretically this code should be fine.

However, if the witness is initialized with nil chain (for stateless execution), then it will panic if `AddBlockHash` is called, although it's not supposed to be called.

Any particular reason to remove the panic protection?

Reply:
You mean why did I remove the check for tail == nil? Because I *always* pass the parent header to the witness on cosntruction because I need the pre-root hash. The original code passed the pre-root separately a a field but I realized I can't validate it, so I need the parent header in full. If I always have that however, then a tail always exists.

Reply:
If you are wondering about chain being nil and subsequently w.chain.GetHeader crashing. Yeah, that's pretty much a programming error I can't recover from. I.e. I can't create a witness without a chain, so being "smart" will at best "not create" a witness and silently fail, so bad. No meaningful way really apart of returning an error, but that's too much devensive programming for my taste.